libsolidity/ast/Types.cpp

```diff
@@ -2146,30 +2147,70 @@ TypePointer StructType::interfaceType(bool _inLibrary) const
 
 bool StructType::canBeUsedExternally(bool _inLibrary) const
 {
-	if (_inLibrary && location() == DataLocation::Storage)
-		return true;
-	else if (recursive())
-		return false;
-	else
+	if (_inLibrary && m_canBeUsedExternally_library.is_initialized())
+		return *m_canBeUsedExternally_library;
+
+	if (!_inLibrary && m_canBeUsedExternally.is_initialized())
+		return *m_canBeUsedExternally;
+
+	bool result = true;
+
+	auto visitor = [&](
+		StructDefinition const& _struct,
+		CycleDetector<StructDefinition>& _cycleDetector,
+		size_t /*_depth*/
+	)
 	{
 		// Check that all members have interface types.
-		// We pass "false" to canBeUsedExternally (_inLibrary), because this struct will be
-		// passed by value and thus the encoding does not differ, but it will disallow
-		// mappings.
-		// Also return false if at least one struct member does not have a type.
+		// Return false if at least one struct member does not have a type.
 		// This might happen, for example, if the type of the member does not exist,
 		// which is reported as an error.
-		for (auto const& var: m_struct.members())
+		for (ASTPointer<VariableDeclaration> const& variable: _struct.members())
 		{
 			// If the struct member does not have a type return false.
 			// A TypeError is expected in this case.
-			if (!var->annotation().type)
-				return false;
-			if (!var->annotation().type->canBeUsedExternally(false))
-				return false;
+			if (!variable->annotation().type)
+			{
+				result = false;
+				return;
+			}
+
+			Type const* memberType = variable->annotation().type.get();
+
+			while (dynamic_cast<ArrayType const*>(memberType))
+				memberType = dynamic_cast<ArrayType const*>(memberType)->baseType().get();
+
+			if (StructType const* innerStruct = dynamic_cast<StructType const*>(memberType))
+				if (_cycleDetector.run(innerStruct->structDefinition()))
+				{
+					if (_inLibrary)
+						continue;
+					else
+					{
+						result = false;
+						return;
+					}
+				}
+
+			if (!memberType->canBeUsedExternally(_inLibrary))
+			{
+				result = false;
+				return;
+			}
 		}
-	}
-	return true;
+	};
+
+	auto cycleResult = CycleDetector<StructDefinition>(visitor).run(structDefinition());
+
+	if (!_inLibrary && cycleResult != nullptr)
+		result = false;
+
+	if (_inLibrary)
+		m_canBeUsedExternally_library = result;
```

Comment:
Can we determine the value for both at the same time? Isn't it basically the same work?

Reply:
Well, not exactly. If it's `!_inLibrary` it stops when it detects a recursion thus the check for internal function types is not run on the whole tree.

Reply:
I suppose in the case of `inLibrary == true` we can set both, though